### PR TITLE
optimize:refactor of auto proxying of datasource.

### PR DESCRIPTION
**Pull request by xingfudeshi**
### Ⅰ. Describe what this PR did
This PR is totally the refactor of auto proxying of datasource.
Add a new annotation @EnableAutoDataSourceProxy.
User can choose to use CGLIB proxy or JDK proxy.
fix known problems,like
1.final class can't  be proxied.
2.com.sun.proxy.$Proxy153 cannot be cast to com.zaxxer.hikari.HikariDataSource
etc.

### Ⅱ. Does this pull request fix one issue?

fixex #2083 #2124 

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
I've done integration testing.

### CODE REVIEW

File: spring/src/main/java/io/seata/spring/annotation/datasource/EnableAutoDataSourceProxy.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+import org.springframework.context.annotation.Import;
+
+/**
+ * @author xingfudeshi@gmail.com
+ * @date 2019/12/26
```
Comment (zjinlei) [inline spring/src/main/java/io/seata/spring/annotation/datasource/EnableAutoDataSourceProxy.java]: remove @date

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+ * @author xingfudeshi@gmail.com
+ * @date 2019/12/26
+ * The type seata data source bean post processor
+ */
+public class SeataDataSourceBeanPostProcessor implements BeanPostProcessor {
+    private static final Logger LOGGER = LoggerFactory.getLogger(SeataDataSourceBeanPostProcessor.class);
+
+    @Override
+    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+        if (bean instanceof DataSource && !(bean instanceof DataSourceProxy) && ConfigurationFactory.getInstance().getBoolean(DATASOURCE_AUTOPROXY, false)) {
```
Comment (zjinlei) [inline spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java]: Use EnableAutoDataSourceProxy, is ConfigurationKeys.DATASOURCE_AUTOPROXY still necessary?

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    /**
+     * proxy data source
+     *
+     * @param originBean
+     * @return proxied datasource
+     */
+    private Object proxyDataSource(Object originBean) {
+        DataSourceProxy dataSourceProxy = DataSourceProxyHolder.get().putDataSource((DataSource) originBean);
+        if (ConfigurationFactory.getInstance().getBoolean(DATASOURCE_USE_JDK_PROXY, false)) {
```
Comment (zjinlei) [inline spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java]: In 1.0.0, jdk proxy is default, recommendations for consistency.

Comment (xingfudeshi) [reply spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java]: Based on the known usage, cglib proxy can be adapted to all third parties except Java melody. In addition, the JDK proxy needs to use interface injection, and there may be some code intrusion for some users' existing business.

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java (hunk 2 above)
Comment (xingfudeshi) [reply spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java]: Probably no longer required.

File: spring/src/main/java/io/seata/spring/annotation/datasource/EnableAutoDataSourceProxy.java (hunk 1 above)
Comment (xingfudeshi) [reply spring/src/main/java/io/seata/spring/annotation/datasource/EnableAutoDataSourceProxy.java]: OK.

File: script/client/conf/file.conf [hunk 4]
```diff
@@ -63,8 +63,4 @@ client {
   log {
     exceptionRate = 100
   }
-  support {
```
Comment (slievrly) [inline script/client/conf/file.conf]: if it not use sb starter, Isn't this switch needed?

Comment (xingfudeshi) [reply script/client/conf/file.conf]: I will remove it.

File: spring/src/main/java/io/seata/spring/annotation/datasource/AutoDataSourceProxyRegistrar.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ * @author xingfudeshi@gmail.com
+ * The type auto data source proxy registrar
+ */
+public class AutoDataSourceProxyRegistrar implements ImportBeanDefinitionRegistrar {
+    private static final String ATTRIBUTE_KEY_USE_JDK_PROXY = "useJdkProxy";
+    public static final String BEAN_NAME_SEATA_DATA_SOURCE_BEAN_POST_PROCESSOR = "seataDataSourceBeanPostProcessor";
+
+    @Override
+    public void registerBeanDefinitions(AnnotationMetadata importingClassMetadata, BeanDefinitionRegistry registry) {
+        if (!registry.containsBeanDefinition(BEAN_NAME_SEATA_DATA_SOURCE_BEAN_POST_PROCESSOR)) {
```
[the remark this answers is not in the document]
Comment (xingfudeshi) [reply spring/src/main/java/io/seata/spring/annotation/datasource/AutoDataSourceProxyRegistrar.java]: Not required.

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+     * @param originBean
+     * @return proxied datasource
+     */
+    private Object proxyDataSource(Object originBean) {
+        DataSourceProxy dataSourceProxy = DataSourceProxyHolder.get().putDataSource((DataSource) originBean);
+        if (this.useJdkProxy) {
+            return Proxy.newProxyInstance(Thread.currentThread().getContextClassLoader(), SpringProxyUtils.getAllInterfaces(originBean), (proxy, method, args) -> handleMethodProxy(dataSourceProxy, method, args, originBean));
+        } else {
+            return Enhancer.create(originBean.getClass(), (MethodInterceptor) (proxy, method, args, methodProxy) -> handleMethodProxy(dataSourceProxy, method, args, originBean));
+        }
```
Comment (XCXCXCXCX) [inline spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java]: return this.useJdkProxy ? jdkProxy() : cglibProxy();

File: spring/src/main/java/io/seata/spring/annotation/datasource/EnableAutoDataSourceProxy.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+@Import(AutoDataSourceProxyRegistrar.class)
+@Documented
+public @interface EnableAutoDataSourceProxy {
+    /**
+     * Whether use JDK proxy instead of CGLIB proxy
+     *
+     * @return useJdkProxy
+     */
+    boolean useJdkProxy() default false;
+}
```
Comment (XCXCXCXCX) [inline spring/src/main/java/io/seata/spring/annotation/datasource/EnableAutoDataSourceProxy.java]: It reminds me to make an annotation EnableAspectJAutoProxy, maybe it should be named after it

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Override
+    public Object postProcessBeforeInitialization(Object bean, String beanName) {
+        if (bean instanceof DataSourceProxy) {
+            throw new ShouldNeverHappenException("Auto proxy of DataSource can't be enabled as you've created a DataSourceProxy bean." +
+                "Please consider removing DataSourceProxy bean or disabling auto proxy of DataSource.");
+        }
+        return bean;
+    }
```
[the remark this answers is not in the document]
Comment (zjinlei) [reply spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java]: > Why throw ex when user injects a proxy bean ? I think skipping the proxy is fine.

DataSource and DataSourceProxy will be proxied separately, and the DataSource held by DataSourceProxy will become proxy class.

File: spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java (hunk 6 above)
Comment (xingfudeshi) [reply spring/src/main/java/io/seata/spring/annotation/datasource/SeataDataSourceBeanPostProcessor.java]: It's not necessary,it's just one line code.I will add separate methods if there are more codes in the future.

File: seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/SeataProperties.java [hunk 9]
```diff
@@ -41,6 +39,14 @@
...
      */
     private String txServiceGroup;
+    /**
+     * Whether enable auto proxying of datasource bean
+     */
+    private boolean enableAutoDataSourceProxy = true;
+    /**
+     * Whether use JDK proxy instead of CGLIB proxy
+     */
+    private boolean useJdkProxy = false;
```
Comment (zjinlei) [inline seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/SeataProperties.java]: add them to yml and properties under script

Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/SeataProperties.java]: OK

Comment (xingfudeshi) [reply seata-spring-boot-starter/src/main/java/io/seata/spring/boot/autoconfigure/properties/SeataProperties.java]: done

**Pull request merged**